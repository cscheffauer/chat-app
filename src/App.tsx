import React, { useState } from 'react';
import { w3cwebsocket as W3CWebSocket } from "websocket";

import './index.scss';
import Homepage from './containers/Homepage/Homepage.component';
import Header from './components/Header/Header.component';


export const client = new W3CWebSocket('ws://localhost:8001');


const App = () => {
  const [connection, setconnection] = useState('loading');

  client.onopen = () => {
    setconnection('established');
    console.log('WebSocket Client Connected');
  };

  client.onerror = () => {
    setconnection('error');
    console.log("Can't connect to Websocket Server.")
  }

  const renderConnectionSwitch = () => {
    switch (connection) {
      case 'loading':
        return <p>Loading...</p>
      case 'established':
        return <Homepage client={client} />
      case 'error':
        return <div className="errorboundary">< h3 >Chat is currently unavailable!</h3 > <p>Pls contact support.</p></div >
      default:
        return <div />
    }
  }

  return (
    <>
      <Header />
      {
        renderConnectionSwitch()
      }
    </>
  );

}

export default App;