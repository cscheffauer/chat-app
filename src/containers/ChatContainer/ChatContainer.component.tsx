import React, { useState, useEffect, Suspense, lazy } from 'react';
import { MessageType } from '../../models/chat.model'
import UserList from '../../components/UserList/UserList.component';

const AsyncChat = lazy(() => import('../../components/Chat/Chat.component'));
const AsyncUserList = lazy(() => import('../../components/UserList/UserList.component'));
const AsyncTabs = lazy(() => import('../../components/Tabs/Tabs.component'));

interface Props {
    username: String,
    client: any
}

const ChatContainer = ({ client, username }: Props) => {
    const [userList, setuserList] = useState({});
    const [userid, setuserid] = useState("");
    const [participantNumber, setparticipantNumber] = useState(0);
    const [messages, setmessages] = useState([] as Array<MessageType>);
    const [selected, setselected] = useState(2);        //set initial tab to be shown to Chat tab (=2)

    useEffect(() => {           //TODO: might not be needed because userlist length can be set directly
        setparticipantNumber(Object.keys(userList).length)      //get keys of userList and set the length of it as the participant number, whenever userlist changes
    }, [userList])

    client.onmessage = (incomingMessage: any) => {
        const dataFromServer = JSON.parse(incomingMessage.data);
        console.log(dataFromServer);

        if (dataFromServer.type === "useridevent") {
            setuserid(dataFromServer.data.userid);      //reacting on incoming user id events and save just the userid
        }
        if (dataFromServer.type === "userevent") {
            setuserList(dataFromServer.data.users);
            setmessages(dataFromServer.data.messages);      //reacting on incoming user events and save the user & messages data
        }
        if (dataFromServer.type === "newmessageevent"
            || dataFromServer.type === "getallmessagesevent"
            || dataFromServer.type === "editmessageevent"
            || dataFromServer.type === "deletemessageevent") {
            setmessages(dataFromServer.data.messages);              //reacting on all incoming message events and save the messages data
        }
    };

    const sendMessage = (text: String) => {
        const sent = new Date();
        const messageJson = { text, username, userid, sent }
        client.send(JSON.stringify({
            message: messageJson,
            type: "newmessageevent"
        }));
    }

    const editMessage = (messageid: String, text: String) => {
        console.log(messageid, text);
        client.send(JSON.stringify({
            id: messageid,
            text: text,
            type: "editmessageevent"
        }));
    }

    const deleteMessage = (messageid: String) => {
        client.send(JSON.stringify({
            id: messageid,
            type: "deletemessageevent"
        }));
    }

    return (
        <div className="chatcontainer">
            <Suspense fallback={< div style={{ textAlign: 'center' }} > Loading...</div >}>
                <AsyncTabs setselected={setselected} selected={selected} participantNumber={Object.keys(userList).length} />
                {selected === 1 ?           //if tabIndex 1 (=User List) is selected
                    <AsyncUserList userList={userList} />
                    :                       //if tabIndex 2 (=Chat) is selected
                    <AsyncChat sendMessage={sendMessage} editMessage={editMessage} deleteMessage={deleteMessage} messages={messages} userid={userid} />
                }
            </Suspense>
        </div>
    )
}

export default ChatContainer;
