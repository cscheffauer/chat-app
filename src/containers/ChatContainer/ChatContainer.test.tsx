import React from 'react';
import { shallow } from 'enzyme';
import ChatContainer from './ChatContainer.component';

import { client } from '../../App';

describe('Container / ChatContainer', () => {
    const clientMock = client;
    const ChatContainerWrapper = shallow(<ChatContainer client={clientMock} username={'John Doe'} />);

    it('expect to render ChatContainer container', () => {
        expect(ChatContainerWrapper).toMatchSnapshot();
    })
})


