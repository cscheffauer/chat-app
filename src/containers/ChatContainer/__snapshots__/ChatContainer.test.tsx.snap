// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container / ChatContainer expect to render ChatContainer container 1`] = `
<div
  className="chatcontainer"
>
  <UserList
    userList={Object {}}
  />
  <Chat
    messages={Array []}
    sendMessage={[Function]}
  />
</div>
`;
